common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -1410,6 +1422,13 @@ public boolean equals(final Object other) {
     }
   }
 
+  /**
+   * Collation-aware equality comparison of two UTF8String.
+   */
+  public boolean semanticEquals(final UTF8String other, int collationId) {
+    return CollationFactory.fetchCollation(collationId).equalsFunction.apply(this, other);
```

Comment:
I'm a bit worried about perf issues here. This is invoked per-row. How much is the overhead here if the collation id is actually `UCS_BASIC`?

Reply:
In theory, impact should be negligible. `fetchCollation` is O(1) operation, so this should only add a couple of extra instructions. Let's keep as is here but add special handling for `UCS_BASIC` in `CodeGen`, as you suggested?

Reply:
It would be nice to add some benchmarks for string ops. @dbatomic Could you re-check the existing benchmarks 
```
find . -name "*Benchmark.scala" |wc -l
      76
```
Maybe there are some relevant benchmarks already.

Reply:
Would it be ok to do benchmarks as a follow up PR? We definitely do need benchmark strategy for both regular string ops and collation special cases.

The closest we have so far in this space is `CharVarcharBenchmark` which is not really what we need.

IMO, this PR is getting a bit too large for my taste so I would prefer to split it into smaller parts. What do you think?

Reply:
> Would it be ok to do benchmarks as a follow up PR?

Let's do that independently from the PR.